### PR TITLE
Throw reasonable error when incompatible temporal unit is used in time unit parameter

### LINKED ISSUE #46068: [BE] When the temporal-unit cannot be applied to a column, return an error message explaining that and not a generic error

### PR DESCRIPTION
**Pull request by lbrdnk**
Closes https://github.com/metabase/metabase/issues/46068

Context [on slack](https://metaboat.slack.com/archives/C0645JP1W81/p1722596802747439).

### CODE REVIEW

File: src/metabase/query_processor/util/temporal_bucket.clj [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  "Returns valid temrpoal units for the `base-type`."
+  {:arglists '([base-type])}
+  keyword)
+
+;; for stuff like UNIX timestamps -- skip validation for now. (UNIX timestamp should be bucketable with any unit
+;; anyway). Once `:effective-type` is in place, we can actually check those Fields here.
+(defmethod valid-units-for-base-type :type/*        [_] valid-datetime-units)
+(defmethod valid-units-for-base-type :type/Date     [_] valid-date-units)
+(defmethod valid-units-for-base-type :type/Time     [_] valid-time-units)
+(defmethod valid-units-for-base-type :type/DateTime [_] valid-datetime-units)
```
Comment (lbrdnk) [inline src/metabase/query_processor/util/temporal_bucket.clj:29]: This has been moved from `validate-temporal-bucketing` ns here as it is now common for multiple middlewares.

File: src/metabase/query_processor/util.clj [hunk 2]
```diff
@@ -227,3 +229,16 @@
...
+  ie. not in its source query or source queries of its joins."
+  [{source-metadatas :source-metadata :as _legacy-query} [_tag id-or-name opts :as _field]]
+  (or (:base-type opts)
+      (when (integer? id-or-name)
+        (:base-type (lib.metadata/field (qp.store/metadata-provider) id-or-name)))
+      (when (string? id-or-name)
+        (some (fn [{:keys [name base_type] :as _source-metadata}]
+                (when (= name id-or-name)
+                  base_type))
+              source-metadatas))))
```
Comment (metamben) [inline src/metabase/query_processor/util.clj:244]: Again I wonder what it means looking for the type in the source metadata. That doesn't seem to account for expressions, for example.

Comment (lbrdnk) [reply src/metabase/query_processor/util.clj:244]: Returning to this PR I've realized that base type is mandatory for string named fields. Hence this function is redundant in this effort. I will modify the code shortly.

_I believe_ that annotate already does that, or not? I assume the `metabase.query-processor.middleware.annotate/col-info-for-expression` is used for that.

**Review by ranquild — APPROVED** [review]

**Review by metamben — APPROVED** [review]

File: test/metabase/query_processor_test/date_bucketing_test.clj [hunk 3]
```diff
@@ -1366,6 +1367,20 @@
       (testing "annual"
         (is (= [488444.41] (unit-totals "year")))))))
 
+(deftest incompatible-temporal-unit-parameter-test
```
Comment (camsaul) [inline test/metabase/query_processor_test/date_bucketing_test.clj:1370]:
```suggestion
(deftest ^:parallel incompatible-temporal-unit-parameter-test
```

**Review by camsaul — APPROVED** [review]
Overall this WFM

**Pull request merged**